ga4gh/versioning.py

```diff
@@ -0,0 +1,152 @@
+"""
+Classes for implementing and enforcing API versioning
+
+Each route belongs to a RouteGroup.
+Each RouteGroup specifies a set of rules which, when satisfied,
+users are allowed to access its routes.  These rules are:
+    1) a lifecycle
+    2) enabled and disabled versions
+
+Right now, three lifecycle stages are defined:
+    development - the default; endpoints that are under active development;
+        only can be accessed in development mode
+    active - live and stable; able to be accessed in any mode
+    deactivated - not able to be accessed in any mode
+
+Enabled and disabled versions:
+    these two versions specify the min and max versions at which the
+    endpoints can be accessed
+"""
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import functools
+
+
+class Lifecycle(object):
+    """
+    A lifecycle stage for a route group
+    """
+    # TODO more lifecycle stages?
```

Comment:
Devil's Advocate:
(1) why do we need explicitly track Deactivated groups? why not just remove support from the server when we deactivate a group, and return some sort of "unsupported" error for invalid versions?
(2) why do we need Development? Is this going to be enforced in code in some way, rather than just through convention and out-of-band communication?

(If so, then one might eventually want to have a BETA state, but I think this is plenty for an initial implementation.)